### ISSUE TITLE
NCC implement error (with test code)

### ISSUE THREAD

**Issue by flealq** [voxelmorph/voxelmorph#227]
@adalca sorry for bothering. but still rather confused by the ncc implementation.
current line 58-65 of voxelmorph/torch/losses.py are:

        58 u_I = I_sum / win_size
        59 u_J = J_sum / win_size
        60 
        61 cross = IJ_sum - u_J * I_sum - u_I * J_sum + u_I * u_J * win_size
        62 I_var = I2_sum - 2 * u_I * I_sum + u_I * u_I * win_size
        63 J_var = J2_sum - 2 * u_J * J_sum + u_J * u_J * win_size
        64 
        65 cc = cross * cross / (I_var * J_var + 1e-5)

# 1. code may wrong
take [58, 59] into [61-63], we have:
cross = IJ_sum - I_sum * J_sum / win_size 
I_var = I2_sum - I_sum * I_sum / win_size 
J_var = J2_sum - J_sum * J_sum / win_size 

this is unresonable even if we only take computation complexity into account. 
suppose two image I and J, their local smoothed version is I_bar and J_bar
then current implementation is
<a href="https://www.codecogs.com/eqnedit.php?latex=cc=\frac{(\overline{I\cdot&space;J}-\overline{I}\cdot&space;\overline{J})^{2}}{(\overline{I^2}-\overline{I}^2)\cdot(\overline{J^2}-\overline{J}^2)&space;}&space;(1)" target="_blank"><img src="https://latex.codecogs.com/gif.latex?cc=\frac{(\overline{I\cdot&space;J}-\overline{I}\cdot&space;\overline{J})^{2}}{(\overline{I^2}-\overline{I}^2)\cdot(\overline{J^2}-\overline{J}^2)&space;}&space;(1)" title="cc=\frac{(\overline{I\cdot J}-\overline{I}\cdot \overline{J})^{2}}{(\overline{I^2}-\overline{I}^2)\cdot(\overline{J^2}-\overline{J}^2) }       (1)" /></a>
**and (1) is definitely not the paper means to be.**

notice, i'd tested (1)'s result is equivalent to line [65], and of course faster.
so i dont think this is a matter of efficiency.

# 2 author may try to
i think line [61] should be at least:
61 cross = IJ_sum - u_J * I - u_I * J + u_I * u_J * win_size
but still, this cross need to conv filtered before the square step: 
cross = conv_fn(IJ_sum - u_J * I - u_I * J + u_I * u_J * win_size)
same questions for line 62 and 63.

# 3 my code
my code is attached below.
and because my code still do not work, i think better leave it to the author.
thank you.

# 4 test code
```python
import torch
import torch.nn.functional as F
from functools import partial
import numpy as np
import math

# the original code, with little revise such as window size and return value
def original_loss(y_true, y_pred):
    I = y_true
    J = y_pred

    # get dimension of volume
    # assumes I, J are sized [batch_size, *vol_shape, nb_feats]
    ndims = len(list(I.size())) - 2
    assert ndims in [1, 2, 3], "volumes should be 1 to 3 dimensions. found: %d" % ndims

    # set window size
    win = [3] * ndims

    # compute filters
    sum_filt = torch.ones([1, 1, *win]).to("cuda")

    pad_no = math.floor(win[0] / 2)

    if ndims == 1:
        stride = (1)
        padding = (pad_no)
    elif ndims == 2:
        stride = (1, 1)
        padding = (pad_no, pad_no)
    else:
        stride = (1, 1, 1)
        padding = (pad_no, pad_no, pad_no)

    # get convolution function
    conv_fn = getattr(F, 'conv%dd' % ndims)

    # compute CC squares
    I2 = I * I
    J2 = J * J
    IJ = I * J

    I_sum = conv_fn(I, sum_filt, stride=stride, padding=padding)
    J_sum = conv_fn(J, sum_filt, stride=stride, padding=padding)
    I2_sum = conv_fn(I2, sum_filt, stride=stride, padding=padding)
    J2_sum = conv_fn(J2, sum_filt, stride=stride, padding=padding)
    IJ_sum = conv_fn(IJ, sum_filt, stride=stride, padding=padding)

    win_size = np.prod(win)
    u_I = I_sum / win_size
    u_J = J_sum / win_size

    cross = IJ_sum - u_J * I_sum - u_I * J_sum + u_I * u_J * win_size
    I_var = I2_sum - 2 * u_I * I_sum + u_I * u_I * win_size
    J_var = J2_sum - 2 * u_J * J_sum + u_J * u_J * win_size

    cc = cross * cross / (I_var * J_var + 1e-10)

    return cross.mean(), I_var.mean(), J_var.mean(), cc.mean()
    # return -torch.mean(cc)


# modify original code for simplicity
# same result with tolerable error
def calc_cross_simple(I, J):
    win = [3, 3]
    sum_filt = torch.ones(1, 1, *win).cuda() / np.prod(win)
    padding = [x // 2 for x in win]
    conv_fn = partial(F.conv2d, weight=sum_filt, stride=1, padding=padding)

    I2 = I * I
    J2 = J * J
    IJ = I * J

    I_sum = conv_fn(I)
    J_sum = conv_fn(J)
    I2_sum = conv_fn(I2)
    J2_sum = conv_fn(J2)
    IJ_sum = conv_fn(IJ)

    win_size = np.prod(win)

    cross = IJ_sum - I_sum * J_sum # err 1e-8
    I_var = I2_sum - I_sum * I_sum # err 1e-8
    J_var = J2_sum - J_sum * J_sum # err 1e-8
    cc = cross * cross / (I_var * J_var + 1e-10) # err 1e-8

    return cross.mean() * win_size, I_var.mean() * win_size, J_var.mean() * win_size, cc.mean()


# my implementation accroding to the paper
def calc_cross_my(I, J):
    win = [3, 3]
    win_size = np.prod(win)
    sum_filt = torch.ones(1, 1, *win).cuda() / np.prod(win)
    padding = [x // 2 for x in win]
    conv_fn = partial(F.conv2d, weight=sum_filt, stride=1, padding=padding)

    I_sum = conv_fn(I)
    J_sum = conv_fn(J)

    cross = conv_fn((I - I_sum) * (J - J_sum)) # totally different result
    I_var = conv_fn((I - I_sum).pow(2))
    J_var = conv_fn((J - J_sum).pow(2))
    cc = cross * cross / (I_var * J_var + 1e-10) # err 1e-8

    return cross.mean() * win_size, I_var.mean() * win_size, J_var.mean() * win_size, cc.mean()


#maybe what the original code want to be?
def original_may_want_to(I, J):
    win = [3, 3]
    win_size = np.prod(win)
    sum_filt = torch.ones(1, 1, *win).cuda() / np.prod(win)
    padding = [x // 2 for x in win]
    conv_fn = partial(F.conv2d, weight=sum_filt, stride=1, padding=padding)

    I2 = I * I
    J2 = J * J
    IJ = I * J

    I_sum = conv_fn(I)
    J_sum = conv_fn(J)

    cross = conv_fn(IJ - I_sum * J - I * J_sum + I_sum * J_sum)
    I_var = conv_fn(I2 - 2 * I * I_sum + I_sum * I_sum)
    J_var = conv_fn(J2 - 2 * J * J_sum + J_sum * J_sum)
    cc = cross * cross / (I_var * J_var + 1e-10)

    return cross.mean() * win_size, I_var.mean() * win_size, J_var.mean() * win_size, cc.mean()


def print_tensor_list(l):
    print([x.detach().cpu().numpy().tolist() for x in l])


def check_ncc():
    torch.torch.manual_seed(0)
    for i in range(10):
        print('test {}'.format(i))
        I = torch.rand([1, 1, 10, 10]).cuda()
        J = torch.rand([1, 1, 10, 10]).cuda()
        # wrong
        org = original_loss(I, J)
        simple = calc_cross_simple(I, J)

        # maybe right?
        my = calc_cross_my(I, J)
        org_rev = original_may_want_to(I, J)

        print_tensor_list(org)
        print_tensor_list(simple)
        print_tensor_list(my)
        print_tensor_list(org_rev)
        debug = 1



if __name__ == '__main__':
    check_ncc()
``` ```

**Comment by flealq** [comment]
move back to #226

**Comment by kvttt** [comment]
```
def calc_cross_my(I, J):
    win = [3, 3]
    win_size = np.prod(win)
    sum_filt = torch.ones(1, 1, *win) / np.prod(win)  # Average filter
    padding = [x // 2 for x in win]
    conv_fn = partial(F.conv2d, weight=sum_filt, stride=1, padding=padding)

    I_sum = conv_fn(I)
    # Each voxel in I_sum represents the average intensity of I around this voxel within the window.
    # i.e. f_hat(p) in the paper.
    J_sum = conv_fn(J)
    # Same as above, but of J. i.e. m_hat . phi (p) in the paper.

    cross = conv_fn((I - I_sum) * (J - J_sum))  # totally different result
    # I - I_sum is essentially f(p) - f_hat(p). Note that in the paper, it is f(p_i) - f_hat(p) instead; 
    # In other words, the paper is suggesting that for each window, you are supposed to subtract f_hat(p) 
    # from EVERY voxel (not just the voxel in the middle) in the window, do the multiplication, 
    # and then take the sum of these (in your case, 9) products.
    I_var = conv_fn((I - I_sum).pow(2))
    # Same issue with the previous one.
    J_var = conv_fn((J - J_sum).pow(2))
    cc = cross * cross / (I_var * J_var + 1e-10) # err 1e-8
    print(cc.size())

    return -torch.mean(cc)
```

@flealq I think the problem has to do with using convolution to calculate the covariance and variance terms. I see that by subtracting I_sum (which in your case should probably be named I_avg) from I directly, you are trying to do the "normalizing" step. But what you have really done there is just left out the voxels in that window other than the one in the middle.

With that being said, I am wondering if calc_cross_my has any sort of mathematical significance given that you mentioned that this loss has been working for you.
